RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,132 @@
+## Usage of `FullyQualifiedName`
+The Test Platform uses `FullyQualifiedName` as a string moniker to name tests during discovery. Visual Studio may filter tests by `FullyQualifiedName` in order to more quickly filter and run specific tests. Visual Studio may also derive type information from the `FullyQualifiedName`. The guidelines below clarify the usage of `FullyQualifiedName` so that these optimized test execution scenarios can be realized.
 
-## Guidelines
-1. **FullyQualifiedName should be unique across your test set.** Internally, TestCase objects are tracked by a unique ID (GUID) which can either be set by the Test Adapter implementor, or automatically generated otherwise. The FullyQualifiedName should be unique if possible, but it's not a requirement.
-1. **FullyQualifiedName should be deterministic** FullyQualifiedName should have the same value for the same test across different discovery sessions. This means that including random data, time-dependent data or environment specific data in the FullyQualifiedName should be avoided.
-1. **FullyQualifiedName is expected to be a reference to the method that implements the test** The Test Platform is expecting the FullyQualifiedName to refer to a particular test case. If there are multiple tests that are instantiated from a test method (data-driven tests for instance), then the FullyQualifiedName should distinguish these using parentheses to represent data parameters and angle brackets to represent type parameters. See below.
+## Background
 
-## Recommended Form
-FullyQualifiedName is expected to be a reference to the TestCase method that implements the test (or tests). The FullyQualifiedName should be unique to a particular TestCase instance, if possible.
+`FullyQualifiedName` (also referenced as FQN) is an identifier supplied by each VS test adapter for each test during discovery. That same FQN value is supplied back to the adapter (along with other TestCase properties) when it's time to execute the tests. One exception to this case is the "Run All" operation which discovers all tests and executes them in one operation. There has historically been no standard for this value and each adapter has implemented its own strategy for creating FQNs.
 
-The recommended forms of FullyQualifiedName are described by the following examples:
+There is a need in VS for each test to have a unique and deterministic method of identifying tests irrespective of the underlying test framework. A standardized FQN opens the path for a full and complete implementation of features like source-based discovery, dynamically fetching source information and UI elements like the hierarchy view in the Test Explorer window.
```

Comment:
For all above features, why can't we let adapters to provide these info? Similar to ITestContainerDiscoverer, we can have one more extensibility point for adapters which can be used by TestWindow to query these info.

Reply:
I'm not sure what you have in mind regarding a new extensibility point. Can you clarify? We are trying to streamline the interface with the adapters by reducing the data and processing required.

Reply:
`NOTE: Our intent is that the test adapters take responsibility for source-based test discovery. Once this transition is complete, the need to keep the FQNs in sync between discovery methods will be reduced to only making sure the types and methods match up.`

Sorry for confusion. My question is related to this section. Why not do this work in first place instead of generalizing FQN?

Instead of generalizing all adapters to follow a unique standard of FQN, why not let adapters make aware of the issues/improvements which they need to do for fast/reliable discovery and execution?
Few of the issues/improvements which adapters needs to be take care of are:
1. Source based discovery
2. Eliminating the need of discovery (even source based discovery) while execution by setting method info while discovery itself (this info can be set in custom properties or in FQN like this PR)

Reply:
This standardization has benefits in making the communication protocol between VS and the adapters more explicit. This has benefits in source-based discovery and execution-by-name as mentioned earlier, but it also creates a foundation for functionality in the future.

This means that we can have a robust foundation for the future, rather than continuing to layer fixes on top of fixes.

Reply:
>This has benefits in source-based discovery and execution-by-name
Thats what my comment is about. In this PR, we are asking adapters to move to standardized FQN so that source based discovery and execution by name can happen properly. Instead we should ask adapters to solve the source based discovery and execution by name problem on their own (its upon adapters to solve this problem via FQN or via some custom property or however they want to).